File: Programmers/LV2_1차_캐시.cpp
#include <string>
#include <vector>
#include <queue>
#include <set>
#include <bits/stdc++.h>

// 8:38

using namespace std;

int solution(int cacheSize, vector<string> cities) {
    int answer = 0;
    list<string> order; // 캐시 순서 확인용
    set<string> cache;  // 캐시 존재 확인용
    
    for(string city : cities) {
        string temp = "";
        for(int i = 0; i < city.size(); i++) temp += tolower(city[i]);
        city = temp;
        if(find(order.begin(), order.end(), city) != order.end()) {
            order.remove(city); order.push_back(city);
            answer += 1; continue;
        }
        else if(cache.size() == cacheSize) { 
            if(cacheSize == 0) { answer += 5; continue; }
            cache.erase(order.front()); cache.insert(city);
            order.remove(order.front()); order.push_back(city);
            answer += 5;
        }
        else {
            order.push_back(city); cache.insert(city);
            answer += 5;
        }  
    }
    return answer;
}

File: Programmers/LV2_1차_캐시.java
import java.lang.*;
import java.util.*;

// 7:44 - 8:08 70
// 8:20 100
class Solution {
    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        List<String> order = new LinkedList<>();
        Set<String> cache = new HashSet<>();
        
        for(String city : cities) {
            city = city.toLowerCase();
            if(cache.contains(city)) {
                answer += 1;
                order.remove(city); order.add(city);
            }
            else if(cache.size() == cacheSize) {
                if(cacheSize == 0) { answer += 5; continue; }
                cache.remove(order.get(0)); cache.add(city);
                order.remove(order.get(0)); order.add(city);
                answer += 5;
            }
            else {
                cache.add(city); order.add(city); 
                answer += 5;
            }
        }
        return answer;
    }
}